File: src/test/com/natpryce/snodge/MutagenAtPathTest.java
package com.natpryce.snodge;

import com.google.gson.JsonElement;
import com.google.gson.JsonPrimitive;
import com.natpryce.snodge.mutagens.ReplaceJsonElement;
import org.junit.Test;

import java.util.List;

import static com.natpryce.snodge.JsonBuilders.object;
import static com.natpryce.snodge.JsonBuilders.withField;
import static com.natpryce.snodge.Mutagens.atPath;
import static java.util.stream.Collectors.toList;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.junit.MatcherAssert.assertThat;
import static org.junit.Assert.assertTrue;

public class MutagenAtPathTest {
    @Test
    public void canLimitMutagenToPath() throws Exception {
        JsonMutator mutator = new JsonMutator(
                atPath(JsonPath.of("a", "b"), new ReplaceJsonElement(new JsonPrimitive("XXX"))));

        JsonElement doc = object(
                withField("a", object(
                        withField("b", 1),
                        withField("c", 2))),
                withField("d", object(
                        withField("b", 1),
                        withField("c", 2)))
        );

        JsonElement mutant = mutator.mutate(doc, 1).findAny().get();

        assertThat(mutant, equalTo((JsonElement) object(
                withField("a", object(
                        withField("b", "XXX"),
                        withField("c", 2))),
                withField("d", object(
                        withField("b", 1),
                        withField("c", 2)))
        )));
    }

    @Test
    public void canLimitMutagenToPathsByPredicate() throws Exception {
        JsonMutator mutator = new JsonMutator(
                atPath(JsonPath.functions.endsWith("b"), new ReplaceJsonElement(new JsonPrimitive("XXX"))));

        JsonElement doc = object(
                withField("a", object(
                        withField("b", 1),
                        withField("c", 2))),
                withField("d", object(
                        withField("b", 1),
                        withField("c", 2)))
        );

        List<JsonElement> mutations = mutator.mutate(doc, 2).collect(toList());

        assertTrue("a.b mutated", mutations.contains(
                object(
                        withField("a", object(
                                withField("b", "XXX"),
                                withField("c", 2))),
                        withField("d", object(
                                withField("b", 1),
                                withField("c", 2)))
                )
        ));

        assertTrue("d.b mutated", mutations.contains(
                object(
                        withField("a", object(
                                withField("b", 1),
                                withField("c", 2))),
                        withField("d", object(
                                withField("b", "XXX"),
                                withField("c", 2)))
                )
        ));
    }
}

File: src/test/com/natpryce/snodge/MutagenAtPathTest.kt
package com.natpryce.snodge

import com.google.gson.JsonElement
import com.google.gson.JsonPrimitive
import com.natpryce.hamkrest.assertion.assertThat
import com.natpryce.hamkrest.equalTo
import com.natpryce.snodge.mutagens.ReplaceJsonElement
import org.junit.Assert.assertTrue
import org.junit.Test

class MutagenAtPathTest {
    @Test
    fun canLimitMutagenToPath() {
        val mutator = JsonMutator(
            ReplaceJsonElement(JsonPrimitive("XXX")).atPath(JsonPath.of("a", "b")))
    
        val doc = obj(
            withField("a", obj(
                withField("b", 1),
                withField("c", 2))),
            withField("d", obj(
                withField("b", 1),
                withField("c", 2))))
        
        val mutant = mutator.invoke(doc, 1).first()
        
        assertThat(mutant, equalTo(obj(
            withField("a", obj(
                withField("b", "XXX"),
                withField("c", 2))),
            withField("d", obj(
                withField("b", 1),
                withField("c", 2)))) as JsonElement))
    }
    
    @Test
    @Throws(Exception::class)
    fun canLimitMutagenToPathsByPredicate() {
        val mutator = JsonMutator(
            ReplaceJsonElement(JsonPrimitive("XXX")).atPath(JsonPath.functions.endsWith("b")))
        
        val doc = obj(
            withField("a", obj(
                withField("b", 1),
                withField("c", 2))),
            withField("d", obj(
                withField("b", 1),
                withField("c", 2))))
        
        val mutations = mutator.invoke(doc, 2)
        
        assertTrue("a.b mutated", mutations.contains(
            obj(
                withField("a", obj(
                    withField("b", "XXX"),
                    withField("c", 2))),
                withField("d", obj(
                    withField("b", 1),
                    withField("c", 2))))))
        
        assertTrue("d.b mutated", mutations.contains(
            obj(
                withField("a", obj(
                    withField("b", 1),
                    withField("c", 2))),
                withField("d", obj(
                    withField("b", "XXX"),
                    withField("c", 2))))))
    }
}
